File: clients/go/aicq/crypto_test.go
package aicq

import (
	"crypto/ed25519"
	"crypto/rand"
	"encoding/base64"
	"testing"
)

func generateTestKeypair(t *testing.T) (ed25519.PrivateKey, string) {
	t.Helper()
	pub, priv, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		t.Fatal(err)
	}
	return priv, base64.StdEncoding.EncodeToString(pub)
}

func TestRoundTrip(t *testing.T) {
	bobPriv, bobPub := generateTestKeypair(t)

	ct, err := EncryptDM("Hello Bob!", bobPub)
	if err != nil {
		t.Fatal(err)
	}
	pt, err := DecryptDM(ct, bobPriv)
	if err != nil {
		t.Fatal(err)
	}
	if pt != "Hello Bob!" {
		t.Fatalf("expected 'Hello Bob!', got %q", pt)
	}
}

func TestWireFormatStructure(t *testing.T) {
	_, priv, _ := ed25519.GenerateKey(rand.Reader)
	pubB64 := base64.StdEncoding.EncodeToString(priv.Public().(ed25519.PublicKey))

	ct, err := EncryptDM("test", pubB64)
	if err != nil {
		t.Fatal(err)
	}
	wire, _ := base64.StdEncoding.DecodeString(ct)
	// 32 (eph pk) + 12 (nonce) + 4 (plaintext) + 16 (tag) = 64
	if len(wire) != 64 {
		t.Fatalf("expected wire length 64, got %d", len(wire))
	}
}

func TestDifferentCiphertexts(t *testing.T) {
	_, priv, _ := ed25519.GenerateKey(rand.Reader)
	pubB64 := base64.StdEncoding.EncodeToString(priv.Public().(ed25519.PublicKey))

	ct1, _ := EncryptDM("same", pubB64)
	ct2, _ := EncryptDM("same", pubB64)
	if ct1 == ct2 {
		t.Fatal("ciphertexts should differ for same plaintext")
	}

	pt1, _ := DecryptDM(ct1, priv)
	pt2, _ := DecryptDM(ct2, priv)
	if pt1 != "same" || pt2 != "same" {
		t.Fatal("both should decrypt to 'same'")
	}
}

func TestWrongKeyFails(t *testing.T) {
	_, priv, _ := ed25519.GenerateKey(rand.Reader)
	pubB64 := base64.StdEncoding.EncodeToString(priv.Public().(ed25519.PublicKey))

	ct, _ := EncryptDM("secret", pubB64)

	_, wrongPriv, _ := ed25519.GenerateKey(rand.Reader)
	_, err := DecryptDM(ct, wrongPriv)
	if err == nil {
		t.Fatal("expected error with wrong key")
	}
	if !ErrCrypto(err) {
		t.Fatalf("expected CryptoError, got %T", err)
	}
}

func TestTamperedCiphertext(t *testing.T) {
	_, priv, _ := ed25519.GenerateKey(rand.Reader)
	pubB64 := base64.StdEncoding.EncodeToString(priv.Public().(ed25519.PublicKey))

	ct, _ := EncryptDM("secret", pubB64)
	wire, _ := base64.StdEncoding.DecodeString(ct)
	wire[len(wire)-1] ^= 0xFF
	tampered := base64.StdEncoding.EncodeToString(wire)

	_, err := DecryptDM(tampered, priv)
	if err == nil {
		t.Fatal("expected error with tampered ciphertext")
	}
}

func TestTruncatedCiphertext(t *testing.T) {
	_, priv, _ := ed25519.GenerateKey(rand.Reader)
	short := base64.StdEncoding.EncodeToString(make([]byte, 30))

	_, err := DecryptDM(short, priv)
	if err == nil {
		t.Fatal("expected error with truncated ciphertext")
	}
}

func TestEmptyPlaintext(t *testing.T) {
	_, priv, _ := ed25519.GenerateKey(rand.Reader)
	pubB64 := base64.StdEncoding.EncodeToString(priv.Public().(ed25519.PublicKey))

	ct, err := EncryptDM("", pubB64)
	if err != nil {
		t.Fatal(err)
	}
	pt, err := DecryptDM(ct, priv)
	if err != nil {
		t.Fatal(err)
	}
	if pt != "" {
		t.Fatalf("expected empty string, got %q", pt)
	}
}

func TestUnicodePlaintext(t *testing.T) {
	_, priv, _ := ed25519.GenerateKey(rand.Reader)
	pubB64 := base64.StdEncoding.EncodeToString(priv.Public().(ed25519.PublicKey))

	msg := "Hello \U0001F30D\u2764\uFE0F \u65E5\u672C\u8A9E"
	ct, err := EncryptDM(msg, pubB64)
	if err != nil {
		t.Fatal(err)
	}
	pt, err := DecryptDM(ct, priv)
	if err != nil {
		t.Fatal(err)
	}
	if pt != msg {
		t.Fatalf("expected %q, got %q", msg, pt)
	}
}

func TestInvalidPublicKeyLength(t *testing.T) {
	_, err := EncryptDM("test", base64.StdEncoding.EncodeToString(make([]byte, 16)))
	if err == nil {
		t.Fatal("expected error with wrong-length key")
	}
	if !ErrCrypto(err) {
		t.Fatalf("expected CryptoError, got %T", err)
	}
}

func TestLargeMessage(t *testing.T) {
	_, priv, _ := ed25519.GenerateKey(rand.Reader)
	pubB64 := base64.StdEncoding.EncodeToString(priv.Public().(ed25519.PublicKey))

	msg := make([]byte, 8000)
	for i := range msg {
		msg[i] = 'A'
	}
	ct, err := EncryptDM(string(msg), pubB64)
	if err != nil {
		t.Fatal(err)
	}
	pt, err := DecryptDM(ct, priv)
	if err != nil {
		t.Fatal(err)
	}
	if pt != string(msg) {
		t.Fatal("large message round-trip failed")
	}
}

func TestBidirectional(t *testing.T) {
	_, alicePriv, _ := ed25519.GenerateKey(rand.Reader)
	alicePub := base64.StdEncoding.EncodeToString(alicePriv.Public().(ed25519.PublicKey))

	_, bobPriv, _ := ed25519.GenerateKey(rand.Reader)
	bobPub := base64.StdEncoding.EncodeToString(bobPriv.Public().(ed25519.PublicKey))

	// Alice -> Bob
	ct1, _ := EncryptDM("Hi Bob", bobPub)
	pt1, err := DecryptDM(ct1, bobPriv)
	if err != nil || pt1 != "Hi Bob" {
		t.Fatal("Alice->Bob failed")
	}

	// Bob -> Alice
	ct2, _ := EncryptDM("Hi Alice", alicePub)
	pt2, err := DecryptDM(ct2, alicePriv)
	if err != nil || pt2 != "Hi Alice" {
		t.Fatal("Bob->Alice failed")
	}
}

File: clients/typescript/src/crypto.test.ts
/**
 * Tests for AICQ DM encryption module.
 * Run: npx ts-node src/crypto.test.ts
 */

import * as assert from "assert";
import * as crypto from "crypto";
import { encryptDM, decryptDM, AICQCryptoError } from "./crypto";

function makeKeypair(): {
  privateKey: crypto.KeyObject;
  publicKeyB64: string;
} {
  const { publicKey, privateKey } = crypto.generateKeyPairSync("ed25519");
  const pubDer = publicKey.export({ type: "spki", format: "der" });
  const pubBytes = (pubDer as Buffer).subarray(-32);
  return {
    privateKey,
    publicKeyB64: pubBytes.toString("base64"),
  };
}

let passed = 0;
let failed = 0;

async function test(name: string, fn: () => Promise<void>): Promise<void> {
  try {
    await fn();
    console.log(`  PASS: ${name}`);
    passed++;
  } catch (e) {
    console.error(`  FAIL: ${name}`);
    console.error(`    ${e}`);
    failed++;
  }
}

async function main() {
  console.log("AICQ Crypto Tests\n");

  const alice = makeKeypair();
  const bob = makeKeypair();

  await test("round-trip encrypt/decrypt", async () => {
    const ct = await encryptDM("Hello Bob!", bob.publicKeyB64);
    const pt = await decryptDM(ct, bob.privateKey);
    assert.strictEqual(pt, "Hello Bob!");
  });

  await test("wire format structure", async () => {
    const plaintext = "test";
    const ct = await encryptDM(plaintext, bob.publicKeyB64);
    const wire = Buffer.from(ct, "base64");
    // 32 (ephemeral pk) + 12 (nonce) + 4 (plaintext) + 16 (tag) = 64
    assert.strictEqual(wire.length, 32 + 12 + plaintext.length + 16);
  });

  await test("different ciphertexts for same plaintext", async () => {
    const ct1 = await encryptDM("same message", bob.publicKeyB64);
    const ct2 = await encryptDM("same message", bob.publicKeyB64);
    assert.notStrictEqual(ct1, ct2);
    assert.strictEqual(await decryptDM(ct1, bob.privateKey), "same message");
    assert.strictEqual(await decryptDM(ct2, bob.privateKey), "same message");
  });

  await test("wrong key fails", async () => {
    const ct = await encryptDM("secret", bob.publicKeyB64);
    await assert.rejects(
      () => decryptDM(ct, alice.privateKey),
      (e: Error) => e instanceof AICQCryptoError && e.message.includes("Decryption failed")
    );
  });

  await test("tampered ciphertext fails", async () => {
    const ct = await encryptDM("secret", bob.publicKeyB64);
    const wire = Buffer.from(ct, "base64");
    wire[wire.length - 1] ^= 0xff;
    const tampered = wire.toString("base64");
    await assert.rejects(
      () => decryptDM(tampered, bob.privateKey),
      AICQCryptoError
    );
  });

  await test("truncated ciphertext fails", async () => {
    const short = Buffer.alloc(30).toString("base64");
    await assert.rejects(
      () => decryptDM(short, bob.privateKey),
      (e: Error) => e instanceof AICQCryptoError && e.message.includes("too short")
    );
  });

  await test("empty plaintext", async () => {
    const ct = await encryptDM("", bob.publicKeyB64);
    const pt = await decryptDM(ct, bob.privateKey);
    assert.strictEqual(pt, "");
  });

  await test("unicode plaintext", async () => {
    const msg = "Hello \u{1F30D}\u{2764}\u{FE0F} \u{65E5}\u{672C}\u{8A9E}";
    const ct = await encryptDM(msg, bob.publicKeyB64);
    const pt = await decryptDM(ct, bob.privateKey);
    assert.strictEqual(pt, msg);
  });

  await test("wrong length public key", async () => {
    const badKey = Buffer.alloc(16).toString("base64");
    await assert.rejects(
      () => encryptDM("test", badKey),
      (e: Error) => e instanceof AICQCryptoError && e.message.includes("Invalid public key length")
    );
  });

  await test("large message", async () => {
    const msg = "A".repeat(8000);
    const ct = await encryptDM(msg, bob.publicKeyB64);
    const pt = await decryptDM(ct, bob.privateKey);
    assert.strictEqual(pt, msg);
  });

  await test("bidirectional", async () => {
    const ct1 = await encryptDM("Hi Bob", bob.publicKeyB64);
    assert.strictEqual(await decryptDM(ct1, bob.privateKey), "Hi Bob");

    const ct2 = await encryptDM("Hi Alice", alice.publicKeyB64);
    assert.strictEqual(await decryptDM(ct2, alice.privateKey), "Hi Alice");
  });

  await test("cross-compatible with Python wire format", async () => {
    // Verify our wire format is: ephemeral_pk[32] + nonce[12] + ciphertext[N+16]
    const ct = await encryptDM("x", bob.publicKeyB64);
    const wire = Buffer.from(ct, "base64");
    // Minimum: 32 + 12 + 1 + 16 = 61 bytes for single-char plaintext
    assert.strictEqual(wire.length, 61);
    // Ephemeral pk should be 32 bytes (valid X25519 point)
    assert.strictEqual(wire.subarray(0, 32).length, 32);
  });

  console.log(`\n${passed} passed, ${failed} failed`);
  if (failed > 0) process.exit(1);
}

main();
